controllers/configurationpolicy_controller.go

```diff
@@ -579,10 +567,33 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
 			}
 		}
 
+		if isNamespaced && len(relevantNamespaces) == 0 {
+			// If relevantNamespaces is empty, retrieve the namespaces based on filters in NamespaceSelector
+			selector := plc.Spec.NamespaceSelector
+			// If MatchLabels/MatchExpressions/Include were not provided, return no namespaces
+			if selector.MatchLabels == nil && selector.MatchExpressions == nil && len(selector.Include) == 0 {
+				log.Info("NamespaceSelector is empty. Skipping namespace retrieval.")
+			} else {
+				relevantNamespaces, err = getPolicyNamespaces(selector)
```

Comment:
This is being called per object template array entry, but this can actually be called only once per `ConfigurationPolicy`. What I suggest is that you determine the namespaces once near the beginning of the `handleObjectTemplates` method as it was before. Optionally, you can only perform the query if at least one object template array entry is of a namespaced object and doesn't specify `metadata.namespace`.

Additionally, it'd be good to add the error to the policy status in the event the namespace selector is invalid. You can do this in a similar way to how the policy template resolve issues get bubbled up.

Reply:
Hopefully the latest changes aren't too sweeping. I've moved the loop that grabs the namespace/name/isNamespaced information up to the top so that the status can be updated if the `NamespaceSelector` fails (if the NamespaceSelector is valid but returns no namespaces, I currently let this fall through to the original "no namespaces" message--maybe it should fail there also?). In the process, the loop stores the namespace/name/isNamespaced information and uses it later in the function rather than re-retrieving it.